File: Java/src/_0016_3Sum_Closest/Solution.java
package _0016_3Sum_Closest;

import java.util.Arrays;

/**
 * 双指针
 */
public class Solution {

    public int threeSumClosest(int[] nums, int target) {

        assert nums.length >= 3;

        // 数组首先排序
        Arrays.sort(nums);

        int diff = Math.abs(nums[0] + nums[1] + nums[2] - target);  // 当前考察值和目标值的差距
        int res = nums[0] + nums[1] + nums[2];  // 当前最好结果

        for (int i = 0; i < nums.length; i++) {
            // 双索引 对于每个nums[i]：不断考察nums[l] 和 nums[r]
            int l = i + 1;
            int r = nums.length - 1;
            int t = target - nums[i];
            while (l < r) {
                if (nums[l] + nums[r] == t)
                    return nums[i] + nums[l] + nums[r];
                else {
                    if (Math.abs(nums[l] + nums[r] - t) < diff) {
                        diff = Math.abs(nums[l] + nums[r] - t);
                        res = nums[i] + nums[l] + nums[r];
                    } else if (nums[l] + nums[r] < t)
                        l++;
                    else
                        r--;
                }
            }
        }

        return res;
    }

    public static void main(String[] args) {

        int[] nums = {-1, 2, 1, -4};
        int target = 1;
        System.out.println((new Solution()).threeSumClosest(nums, target));
    }
}

File: Python/_0016_3Sum_Closest/Solution.py
# coding=utf-8
from typing import List


# 执行用时：684 ms, 在所有 Python3 提交中击败了5.22%的用户
# 内存消耗：14.8 MB, 在所有 Python3 提交中击败了79.01%的用户
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        res = [1e9, 1e9]
        for i in range(len(nums)):
            j = i + 1; k = len(nums) - 1
            while j < k:
                while j < k - 1 and nums[i] + nums[j] + nums[k - 1] >= target:
                    k -= 1
                s = nums[i] + nums[j] + nums[k]
                res = min(res, [abs(s - target), s])
                if j < k - 1:
                    s = nums[i] + nums[j] + nums[k - 1]
                    res = min(res, [abs(s - target), s])
                j += 1
        return int(res[1])


if __name__ == "__main__":
    print(Solution().threeSumClosest([-1, 2, 1, -4], 1))  # 2
    print(Solution().threeSumClosest([1, 1, -1, -1, 3], 3))  # 3
